src/platform/audio/ffmpegaudiofilecache.cpp

```diff
@@ -0,0 +1,537 @@
+/**
+ * @file
+ *
+ * @author feliwir
+ *
+ * @brief Class for caching loaded audio samples to reduce file IO.
+ *
+ * @copyright Thyme is free software: you can redistribute it and/or
+ *            modify it under the terms of the GNU General Public License
+ *            as published by the Free Software Foundation, either version
+ *            2 of the License, or (at your option) any later version.
+ *            A full copy of the GNU General Public License can be found in
+ *            LICENSE
+ */
+
+extern "C" {
+#include <libavcodec/avcodec.h>
+#include <libavformat/avformat.h>
+}
+
+#include "audioeventrts.h"
+#include "audiomanager.h"
+#include "ffmpegaudiofilecache.h"
+#include "filesystem.h"
+#include <captainslog.h>
+#include <list>
+
+namespace Thyme
+{
+struct WavHeader
+{
+    uint8_t riff_id[4] = { 'R', 'I', 'F', 'F' };
+    uint32_t chunk_size;
+    uint8_t wave_id[4] = { 'W', 'A', 'V', 'E' };
+    /* "fmt" sub-chunk */
+    uint8_t fmt_id[4] = { 'f', 'm', 't', ' ' }; // FMT header
+    uint32_t subchunk1_size = 16; // Size of the fmt chunk
+    uint16_t audio_format = 1; // Audio format 1=PCM
+    uint16_t channels = 1; // Number of channels 1=Mono 2=Sterio
+    uint32_t samples_per_sec = 16000; // Sampling Frequency in Hz
+    uint32_t bytes_per_sec = 16000 * 2; // bytes per second
+    uint16_t block_align = 2; // 2=16-bit mono, 4=16-bit stereo
+    uint16_t bits_per_sample = 16; // Number of bits per sample
+    /* "data" sub-chunk */
+    uint8_t subchunk2_id[4] = { 'd', 'a', 't', 'a' }; // "data"  string
+    uint32_t subchunk2_size; // Sampled data length
+};
+
+FFmpegAudioFileCache::~FFmpegAudioFileCache()
+{
+    ScopedMutexClass lock(&m_mutex);
+
+    for (auto it = m_cacheMap.begin(); it != m_cacheMap.end(); ++it) {
+        Release_Open_Audio(&it->second);
+    }
+}
+
+/**
+ * Read an FFmpeg packet from file
+ */
+int FFmpegAudioFileCache::Read_FFmpeg_Packet(void *opaque, uint8_t *buf, int buf_size)
+{
+    File *file = static_cast<File *>(opaque);
+    int read = file->Read(buf, buf_size);
+
+    // Streaming protocol requires us to return real errors - when we read less equal 0 we're at EOF
+    if (read <= 0)
+        return AVERROR_EOF;
+
+    return read;
+}
+
+/**
+ * Open all the required FFmpeg handles for a required file.
+ */
+bool FFmpegAudioFileCache::Open_FFmpeg_Contexts(FFmpegOpenAudioFile *open_audio, File *file)
+{
+#if LOGGING_LEVEL != LOGLEVEL_NONE
+    av_log_set_level(AV_LOG_INFO);
+#endif
+
+// This is required for FFmpeg older than 4.0 -> deprecated afterwards though
+#if LIBAVFORMAT_VERSION_MAJOR < 58
+    av_register_all();
+#endif
+
+    // FFmpeg setup
+    int result = 0;
+    char error_buffer[1024];
```

Comment:
I suggest to move the stack allocations just before each `av_strerror` call.
Same in next function.